hphp/runtime/vm/jit/vasm-map-flags.cpp

```diff
@@ -0,0 +1,204 @@
+#undef I
+#undef fl
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/*
+ * For each use of an SF, return the StatusFlags bits needed by the use's CC.
+ *
+ * Specializations are below, which can additionally change the CC if the
+ * platform doesn't have a simple CC mapping for the given def and use.
+ */
+template<typename Inst>
+uint8_t mapFlags(Env& env, Inst& inst) {
+  return static_cast<uint8_t>(required_flags(inst.cc));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+uint8_t mapFlags(Env& env, jcc& inst) {
+  if (env.sf_defs[inst.sf]->op == Vinstr::ucomisd) {
+    switch (inst.cc) {
+    case CC_AE: // CC_NB
+      inst.cc = CC_GE;
+      break;
+    default:
+      break;
+    }
+  }
+  return static_cast<uint8_t>(required_flags(inst.cc));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/*
+ * Map this uses CC in any platform-specific way and return the platform-
+ * specific status flags bits that are used for the CC.
+ */
+uint8_t mapFlags(Env& env, Vinstr* inst) {
+  switch (inst->op) {
+#define O(name, imms, ...) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+  }
+#define Inone
+#define cc cc; return mapFlags(env, i)
+#define I(x) (void)i.x;
+    VASM_OPCODES
+#undef I
+#undef cc
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+}
+
+/*
+ * Map SF pass for a Vunit.
+ */
+void mapFlags(Vunit& unit) {
+  assertx(check(unit));
+  auto& blocks = unit.blocks;
+
+  Env env { unit };
+  env.sf_defs.clear();
+  env.sf_uses.clear();
+
+  auto const labels = sortBlocks(unit);
+
+  // Set up Env, only visiting reachable blocks.
+  // Tracks each VregSF to the instruction which def'd it.
+  // Tracks each VregSF to a list of each instruction which uses it.
+  for (auto const b : labels) {
+    assertx(!blocks[b].code.empty());
+    for (auto const& inst : blocks[b].code) {
+      visitDefs(unit, inst, [&] (Vreg r, Width w) {
+        if (w == Width::Flags) {
+          env.sf_defs[r] = (Vinstr*)(&inst);
+        }
+      });
+      visitUses(unit, inst, [&] (Vreg r, Width w) {
+        if (w == Width::Flags) {
+          env.sf_uses[r].push_back((Vinstr*)(&inst));
+        }
+      });
+    }
+  };
+
+  // Calls mapFlags on each 
+  for (auto it = env.sf_defs.begin(); it != env.sf_defs.end(); it++) {
+    auto r = it->first;
+    auto d = it->second;
+
+    // For each instruction that defs an sf, initially clear the flag bits.
+    uint8_t flags = 0x00;
+
+    // Call mapFlags for each use. This will a) update the instruction's
+    // condition code (cc) based on any platform specific mapping b) update
+    // the flag bits (c) for this VregSF def based on the instruction's cc.
+    for (auto const u : env.sf_uses[r]) {
```

Comment:
@dave-estes—I don't see how that could fit into the scope of this pass.  That sounds like a peephole optimization you could do as part of `vasm-simplify.cpp` after running `annotateSFUses()`, because it also depends on the defs of those `cmovq{}`s being functionally chained, and not used as srcs by any later instructions.  This pass should be restricted to annotating flags and conditions.